src/Features/CSharp/Portable/StackTraceExplorer/CSharpStackTraceExplorerService.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.StackTraceExplorer;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.StackTraceExplorer
+{
+    [Shared]
+    [ExportLanguageService(typeof(IStackTraceExplorerService), language: LanguageNames.CSharp)]
+    internal class CSharpStackTraceExplorerService : IStackTraceExplorerService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpStackTraceExplorerService()
+        {
+        }
+
+        public string GetClassMetadataName(string className)
+        {
+            var typeNameSyntax = SyntaxFactory.ParseTypeName(className);
+            if (typeNameSyntax is QualifiedNameSyntax qualifiedNameSyntax)
+            {
+                return GetMetadataNameOfQualifiedNameSyntax(qualifiedNameSyntax);
+            }
+
+            return GetGenericNameWithArity(typeNameSyntax);
+        }
+
+        private static string GetMetadataNameOfQualifiedNameSyntax(QualifiedNameSyntax originalSyntax)
+        {
+            SyntaxNode? syntax = originalSyntax;
+            Stack<string> parts = new();
```

Comment:
not sure why we need a stack, as opposed to a list.  also, not sure why we need a collection at all.  can this not be a simple pooled StringBuilder?

Reply:
The items are added right to left, because that's how the syntax ends up parsing. 

`My.NamespaceName.With.Class`

becomes: 

```
[My.NamespaceName.With].[Class]
               ^-- Left  ^-- Right
[My.NamespaceName].[With]
           ^-- Left   ^-- Right
           
and so on...
```